### PR TITLE
Incoming scr preprocessor

### PR DESCRIPTION
**Pull request by mariusmihaic**
## Reasoning behind the pull request
- Adapted existing scr pre processor to work for sovereign incoming scrs. This is for now mostly **copy-paste** code from existing scr pre processor, only adapted not to use any gas for `ProcessMiniBlock` and `ProcessBlockTransactions`. 
  
## Proposed changes
- Removed old sovereign tx pre processor and replaced with scr pre processor

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: process/block/preprocess/sovereignChainIncomingSCR.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+			break
+		}
+
+		scr.saveAccountBalanceForAddress(miniBlockScrs[txIndex].GetRcvAddr())
+
+		snapshot := scr.handleProcessTransactionInit(preProcessorExecutionInfoHandler, miniBlockTxHashes[txIndex])
+		_, err = scr.scrProcessor.ProcessSmartContractResult(miniBlockScrs[txIndex])
+		if err != nil {
+			scr.handleProcessSovereignSCRError(preProcessorExecutionInfoHandler, snapshot, miniBlockTxHashes[txIndex])
+			break
```
Comment (sasurobert) [inline process/block/preprocess/sovereignChainIncomingSCR.go:159]: return error here.

File: process/block/preprocess/sovereignChainIncomingSCR.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+	scr.blockSizeComputation.AddNumMiniBlocks(1)
+	scr.blockSizeComputation.AddNumTxs(len(miniBlock.TxHashes))
+
+	return nil, txIndex - 1, false, err
+}
+
+func (scr *sovereignChainIncomingSCR) handleProcessSovereignSCRError(preProcessorExecutionInfoHandler process.PreProcessorExecutionInfoHandler, snapshot int, txHash []byte) {
+	errRevert := scr.accounts.RevertToSnapshot(snapshot)
+	if errRevert != nil {
+		log.Debug("scr.sovereignChainIncomingSCR: RevertToSnapshot", "error", errRevert.Error())
```
Comment (sasurobert) [inline process/block/preprocess/sovereignChainIncomingSCR.go:187]: return error here.

File: process/block/preprocess/sovereignChainIncomingSCR.go [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+	"github.com/multiversx/mx-chain-core-go/data/smartContractResult"
+	"github.com/multiversx/mx-chain-go/process"
+)
+
+type sovereignChainIncomingSCR struct {
+	*smartContractResults
+}
+
+func onRequestIncomingSCR(_ uint32, txHashes [][]byte) {
+	log.Warn("sovereignChainIncomingSCR.onRequestIncomingSCR was called; not implemented", "missing scrs hashes", txHashes)
```
Comment (sasurobert) [inline process/block/preprocess/sovereignChainIncomingSCR.go:16]: log.error

File: process/block/preprocess/sovereignChainIncomingSCR.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+func NewSovereignChainIncomingSCR(scr *smartContractResults) (*sovereignChainIncomingSCR, error) {
```
Comment (sasurobert) [inline process/block/preprocess/sovereignChainIncomingSCR.go:19]: add comment.

File: process/block/preprocess/sovereignChainIncomingSCR.go [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+		}
+
+		if err != nil {
+			break
+		}
+
+		scr.saveAccountBalanceForAddress(miniBlockScrs[txIndex].GetRcvAddr())
+
+		snapshot := scr.handleProcessTransactionInit(preProcessorExecutionInfoHandler, miniBlockTxHashes[txIndex])
+		_, err = scr.scrProcessor.ProcessSmartContractResult(miniBlockScrs[txIndex])
```
Comment (sasurobert) [inline process/block/preprocess/sovereignChainIncomingSCR.go:156]: We have to go through the full consensus - but we do not need to execute the SCRs here. As far as I see from consensusV2 this processMiniBlock is called on create block. So we have to add the needed SCRs to be executed in the local map and only that, nothing more.

File: process/block/preprocess/sovereignChainIncomingSCR.go [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+
+		}
+
+		createdMBs = append(createdMBs, miniBlock)
+	}
+
+	return createdMBs, nil
+}
+
+func (scr *sovereignChainIncomingSCR) ProcessMiniBlock(
```
Comment (sasurobert) [inline process/block/preprocess/sovereignChainIncomingSCR.go:105]: add comment to the block. We need to rewrite the names of a set of functions - as they are confusing.

File: process/block/preprocess/sovereignChainIncomingSCR.go (hunk 1 above)
Comment (mariusmihaic) [reply process/block/preprocess/sovereignChainIncomingSCR.go:159]: Right, returned error

File: process/block/preprocess/sovereignChainIncomingSCR.go (hunk 3 above)
Comment (mariusmihaic) [reply process/block/preprocess/sovereignChainIncomingSCR.go:16]: Changed it to error

File: process/block/preprocess/sovereignChainIncomingSCR.go (hunk 4 above)
Comment (mariusmihaic) [reply process/block/preprocess/sovereignChainIncomingSCR.go:19]: Added comment

File: process/block/preprocess/sovereignChainIncomingSCR.go (hunk 5 above)
Comment (mariusmihaic) [reply process/block/preprocess/sovereignChainIncomingSCR.go:156]: Looking through the code it seems like indded on v2 we create the miniblocks using `ProcessMiniBlock` and the processing of the received block is done in `ProcessBlockTransactions`.

I've removedit from here

File: process/block/preprocess/sovereignChainIncomingSCR.go (hunk 6 above)
Comment (mariusmihaic) [reply process/block/preprocess/sovereignChainIncomingSCR.go:105]: Right. We could rename them in another task.
Added comment for now

**Review by raduchis — CHANGES_REQUESTED** [review]

File: process/block/preprocess/sovereignChainIncomingSCR.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+				log.Warn("missing transaction in ProcessBlockTransactions ", "type", miniBlock.Type, "txHash", txHash)
+				return nil, process.ErrMissingTransaction
+			}
+
+			currScr, ok := txInfoFromMap.tx.(*smartContractResult.SmartContractResult)
+			if !ok {
+				return nil, process.ErrWrongTypeAssertion
+			}
+
+			err = scr.saveAccountBalanceForAddress(currScr.GetRcvAddr())
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/block/preprocess/sovereignChainIncomingSCR.go:91]: Keeping consistency with original code. This loads the current balance before processing into `BalanceComputationHandler` component and checks for nil node from db error beforehand

File: process/block/preprocess/sovereignChainIncomingSCR.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+	txShardInfoToSet := &txShardInfo{senderShardID: miniBlock.SenderShardID, receiverShardID: miniBlock.ReceiverShardID}
+
+	scr.scrForBlock.mutTxsForBlock.Lock()
+	for index, txHash := range miniBlockTxHashes {
+		scr.scrForBlock.txHashAndInfo[string(txHash)] = &txInfo{tx: miniBlockScrs[index], txShardInfo: txShardInfoToSet}
+	}
+	scr.scrForBlock.mutTxsForBlock.Unlock()
+
+	scr.blockSizeComputation.AddNumMiniBlocks(1)
+	scr.blockSizeComputation.AddNumTxs(len(miniBlock.TxHashes))
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/block/preprocess/sovereignChainIncomingSCR.go:181]: This is to keep consistency with original code, so I guess it should be fine

File: process/block/preprocess/sovereignChainIncomingSCR.go
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/preprocess/sovereignChainIncomingSCR.go:1]: We had several discussions about it when we started sovereign. We started in a new repo first - but it was just too much rework or copy paste code to be done. It deemed better to use internal pointers and make implementation in functions where it mattered.

**Review by sasurobert — APPROVED** [review]

File: process/block/preprocess/sovereignChainIncomingSCR.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply process/block/preprocess/sovereignChainIncomingSCR.go:91]: ok

File: process/block/preprocess/sovereignChainIncomingSCR.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply process/block/preprocess/sovereignChainIncomingSCR.go:181]: ok

**Review by raduchis — APPROVED** [review]

**Pull request merged**